### DISCUSSION TITLE
TODOs

### DISCUSSION THREAD

**Discussion by mattatz** [Nodi3d/nodi-v2#3] [Ideas]
A miscellaneous list of current TODOs, ranging from large to small tasks.

- [x] Descriptions for nodes
- [x] Feature to make the number of inputs and outputs of a node variable
- [x] Expression node
- [ ] Feature to develop custom nodes
- [ ] Feature to group nodes
- [x] Feature to import geometry data (dxf, stl, obj, fbx, support for other formats will be considered based on requests)
- [x] Nestable node graphs (creating sub-graph nodes that contain a node graph)
- [ ] Provide idioms for enabling iterative processing, similar to Houdini's Block Begin/End nodes.
- [ ] Customizable materials with support for custom shaders
- [ ] Clipping view
- [x] Trimmed surface functions
- [ ] CSG for curves
- [x] CSG for meshes
- [ ] CSG for BReps
- [ ] Offset & fillet operations for curves
- [ ] FRep
- [ ] CAM
- [ ] CAE
- [x] Features for data storage on the cloud and other web service functionalities
- [x] Viewer mode
- [ ] Features to edit UI (UI layout, colors)
- [ ] Features to display data in the viewport (e.g. calculation results related to geometry, such as surface area and volume.)
- [ ] Enable to snap in node editor for edit

**Comment by sorashin** [comment]
node shortcut like Grasshopper

0<4<100 -> Number Slider(min:0 max:100 step:1)
// + string -> Panel

https://www.grasshopper3d.com/forum/topics/popup-shortcuts

**Reply by mattatz** [reply]
@sorashin 

Thank you for your feature request!  
You can now use shortcuts directly from the node addition input field, as shown here.

https://github.com/user-attachments/assets/c541e2c5-4a50-48e7-8d81-9d8d1bef1e45
